### PR TITLE
Mega Menu

### PR DESCRIPTION
**Pull request by sebworks**
- Adding Mega Menu

## Changes
- Renamed  'includes/nav-items-toplevel.html' to '_includes/primary-nav-items.html'  better reflect purpose.
- Added  '_includes/templates/nav-about-us-media.html' to store nav media markup.
- Updated ' _includes/templates/primary-nav.html' to include new mega menu design.
- Added '_includes/templates/sub-nav.html' to separate markup from primary nav and add new mega menu design.
- Updated 'src/static/css/header.less' to remove primary nav styles.
- Added 'src/static/css/nav-primary.less' to store primary nav and sub-nav styes.
- Updated 'src/static/js/modules/desktop-menu-transitions.js' to add event listener for trigger click event.
- Added 'static/img/nav-about-us-video.png' for nav media markup.

## Test
- Visit 'http://localhost:7000/' and hover on 'About'

## Screenshots

Desktop:

screen shot 2015-06-26 at 5 45 50 pm

Tablet:

screen shot 2015-06-26 at 5 46 40 pm

## Review
@jimmynotjim
@anselmbradford
@KimberlyMunoz
@Scotchester

### CODE REVIEW

**Comment by jimmynotjim** [comment]
If you start out narrow and then go wide (like rotating a tablet) the links aren't in 3 columns like they should be.

screen shot 2015-06-29 at 5 06 27 pm

File: _includes/templates/nav-about-us-media.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<a href="/the-bureau/">
+    <img src="../../static/img/nav-about-us-video.png"
```
Comment (jimmynotjim) [inline _includes/templates/nav-about-us-media.html]: I don't think we should use the play arrow overlay if the video isn't going to play in the current window.

File: _includes/templates/primary-nav.html [hunk 2]
```diff
@@ -1,40 +1,20 @@
...
-                            {%- for href, id, caption in first_level %}
-                            <li class="list_item list-expanding_list-item">
-                                <a class="list_link" href="{{ href|e }}">{{ caption|e }}</a>
-                            </li>
-                            {%- endfor %}
-                        </ul>
-                    </div>
-                </div>
+                {%- if sub_nav_items|length > 0 %}
+                    {% include '/templates/sub-nav.html' %}
```
Comment (jimmynotjim) [inline _includes/templates/primary-nav.html]: Instead of a template, should this be a macro?

File: src/static/css/nav-primary.less [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+.primary-nav .list_link {
+    .u-link__colors(@black, @green);
+    .webfont-regular();
+}
+
+
+/* Sliding nav for portrait tablets and smaller
+   ========================================================================== */
+
+.respond-to-max(800px, {
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: This will need to be updated once #676 is merged

File: src/static/css/nav-primary.less [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+        padding-right: 0;
+    }
+
+    .sub-nav_menu {
+        border:none !important;
+    }
+
+    .sub-nav_menu-list {
+        width: 100%;
+
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: Remove extra line break

File: src/static/css/nav-primary.less [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+    // Remove extra padding added by .wrapper__match-content so the nav button
+    // is flush left with the screen.
+    .primary-nav_header-space.wrapper__match-content {
+        padding-left: 0;
+        padding-right: 0;
+    }
+
+    .sub-nav_menu {
+        border:none !important;
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: Where is this set? Any way we can condition it off so we don't need `!important`?

File: src/static/css/nav-primary.less [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+    }
+
+    .sub-nav_menu-list {
+        width: 100%;
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: Shouldn't this automatically be 100% by default?

File: src/static/css/nav-primary.less [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+            display: block;
+            padding: (@grid_gutter-width / 2)
+                     @grid_gutter-width
+                     (@grid_gutter-width / 2)
+                     (@grid_gutter-width / 2);
+            position: relative;
+            background: @gray-20;
+            font-size: unit(18px / @base-font-size-px, em);
+        }
+        & > .list_item > a:hover {
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: Inconsistent line breaks throughout.

File: src/static/css/nav-primary.less [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+
+    // This overlays the main content and when clicked closes the menu.
+    .sliding-nav_page-overlay {
+        position: relative;
+    }
+    .sliding-nav_page-overlay:before {
+        position: absolute;
+        z-index: 999;
+        // background: @black;
+        opacity: 0;
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: Since the background and opacity have been removed here and when it's open, we can remove the related styles.

File: src/static/css/nav-primary.less [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+    .sliding-nav_page {
+        box-sizing: border-box;
+        width: 100%;
+        float: left;
+        background: @white;
+        transition: .2s margin-right ease-in, .2s box-shadow ease-in;
+    }
+    .is-open .sliding-nav_page {
+        border-left: 1px solid @gray-80;
+        margin-right: -80%;
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: Can we use translate instead of margin? It animates smoother.

File: src/static/css/nav-primary.less [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+    }
+
+    .sliding-nav_page.is-scroll-disabled {
+        position: fixed;
+        right: 0;
+        -webkit-transition: .2s right ease-in;
+    }
+    .is-open .sliding-nav_page.is-scroll-disabled {
+        position: fixed;
+        right: -80%;
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: Same here on the translate question above.

File: src/static/css/nav-primary.less [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+   ========================================================================== */
+
+.respond-to-min(801px, {
+
+    .u-show-on-desktop {
+        display: block;
+    }
+
+    .content-line__nav {
+          border: 1px solid @gray-50;
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: Remove extra indentation

File: src/static/css/nav-primary.less [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+    .sub-nav_wrapper {
+        background: @gray-5;
+        width: 100%;
+        overflow: hidden;
+        position: absolute;
+        max-height: 0;
+        left: 0;
+        z-index: 100;
+        // Maybe this will kick on the gpu?
+        -webkit-transform: translate3d(0,0,0);
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: No need for prefixes, we have auto-prefixer that will add them.

File: src/static/css/nav-primary.less [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+
+    .sub-nav_title {
+        .webfont-regular();
+        font-size: 1em;
+        color: @pacific;
+    }
+
+});
+
+.respond-to-range(801px, 814px, {
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: 13px range? Is that correct? Can we do something to remove this?

File: src/static/css/nav-primary.less [hunk 14]
```diff
@@ -0,0 +1,487 @@
...
+                                    how we are working for American consumers.
+                                </div>
+                            </aside>
+                        </div>
+                    </div>
+                </li>
+            </ul>
+        </nav>
+      notes:
+        -
```
Comment (anselmbradford) [inline src/static/css/nav-primary.less]: `notes:` can removed because it doesn't contain any content.

File: src/static/css/nav-primary.less [hunk 15]
```diff
@@ -0,0 +1,487 @@
...
+
+.respond-to-max(800px, {
+
+    html,
+    body {
+        height: 100%;
+    }
+
+    .u-show-on-desktop {
+        display: none;
```
Comment (anselmbradford) [inline src/static/css/nav-primary.less]: Doesn't this need a media query? If it's named `u-show-on-desktop` but hides content at all sizes that could be confusing.

Comment (anselmbradford) [inline src/static/css/nav-primary.less]: nm, realized this whole block is wrapped in a media query further up.

File: src/static/css/nav-primary.less [hunk 16]
```diff
@@ -0,0 +1,487 @@
...
+        opacity: 0;
+        -webkit-transition: .2s opacity ease-in;
+    }
+    .is-open .sliding-nav_page-overlay:before {
+        content: "";
+        top: 0;
+        right: 0;
+        bottom: 0;
+        left: 0;
+        // opacity: 0.7;
```
Comment (anselmbradford) [inline src/static/css/nav-primary.less]: Remove commented out bits here and L:230 before merging.

File: src/static/css/nav-primary.less [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+
+    .primary-nav_top-level-list {
+        padding-left: @grid_gutter-width / 2;
+        margin-bottom: 0;
+        position: static;
+        overflow: visible;
+    }
+
+    .primary-nav_top-level-list.list__horizontal > .list_item {
+        display: inline-block;
```
Comment (anselmbradford) [inline src/static/css/nav-primary.less]: Possibly use CF mixin `u-inline-block();` for IE hacks.

File: src/static/css/nav-primary.less [hunk 18]
```diff
@@ -0,0 +1,487 @@
...
+                            @green, @green, @green, @green, @green);
+            cursor: default;
+        }
+
+        .webfont-regular();
+    }
+
+    .sub-nav_menu-list .list_item {
+        display: block;
+        white-space:nowrap;
```
Comment (anselmbradford) [inline src/static/css/nav-primary.less]: Trivial: Add space after colon.

File: _includes/templates/nav-about-us-media.html (hunk 1 above)
Comment (sebworks) [inline _includes/templates/nav-about-us-media.html]: So I'm going to leave as is but I'll add the capability to auto play the video.

File: src/static/css/nav-primary.less (hunk 8 above)
Comment (sebworks) [inline src/static/css/nav-primary.less]: So this goes back to https://github.com/cfpb/cfgov-refresh/pull/151. Looks like the markup was never added. I'm going to comment out the whole block for now.

File: _includes/templates/nav-about-us-media.html (hunk 1 above)
Comment (Scotchester) [inline _includes/templates/nav-about-us-media.html]: Please do not let anything with sound autoplay, ever.

Comment (sebworks) [inline _includes/templates/nav-about-us-media.html]: It would be the same functionality we have implemented on the bureau page when you click the play button.

Comment (Scotchester) [inline _includes/templates/nav-about-us-media.html]: Ah, sorry, I misunderstood the situation. I just saw "autoplay" and my canine instincts kicked in ;)

Comment (sebworks) [inline _includes/templates/nav-about-us-media.html]: lol, no worries. .  @Scotchester, did you guys ever plan on implementing the 'sliding-nav_page-overlay' markup mentioned here: https://github.com/cfpb/cfgov-refresh/pull/151?

Comment (Scotchester) [inline _includes/templates/nav-about-us-media.html]: Yes, we did intend to, though it probably deserves another conversation now that the design and UX team has shifted significantly.

Comment (sebworks) [inline _includes/templates/nav-about-us-media.html]: thanks dude.

File: src/static/css/nav-primary.less (hunk 13 above)
Comment (sebworks) [inline src/static/css/nav-primary.less]: So it was added in https://github.com/cfpb/cfgov-refresh/commit/30e3484920f4d314d116f31943740a04066229b3. @anselmbradford , can you provide background?

Comment (anselmbradford) [inline src/static/css/nav-primary.less]: For the current menu the links are arranged on one line. When they wrap the second line of links was too close to the first line. This bit added padding to the bottom of the first line for the gap between where the links were on one line and where the menu turned into the tablet format. If this is at all applicable in the new menu layout, perhaps to eliminate it the menu would convert to the tablet layout 13px sooner.

Comment (sebworks) [inline src/static/css/nav-primary.less]: thanks

Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: I'd agree on the 13px sooner, seems silly to have such a trivial breakpoint range.

Comment (sebworks) [inline src/static/css/nav-primary.less]: ok, will change it.

Comment (anselmbradford) [inline src/static/css/nav-primary.less]: It's a hackish approach anyway, if the webfonts didn't load the width of the links would differ. Will we even have one row of links in the new menu anyway? If we will and you think of a better way to determine if they've broken to a second line, go with that.

Comment (sebworks) [inline src/static/css/nav-primary.less]: yep, I'm going to just remove it for now. I think we may be able to use flex-box in the future to solve the issue.

**Comment by sebworks** [comment]
@jimmynotjim, please review this when you get a chance.

File: _includes/templates/nav/about-us-media.html [hunk 19]
```diff
@@ -0,0 +1,11 @@
+<a href="/the-bureau/">
+    <img src="../../static/img/nav-about-us-video.png"
+         alt="About Us video image" />
+    <h4 class="short-desc u-mt15">
+      The CFPB: Working for you.
+    </h4>
+</a>
+<div class="short-desc">
+    This short video covers what the CFPB is and
```
Comment (jimmynotjim) [inline _includes/templates/nav/about-us-media.html]: This could be a better call to action, something like "Watch our short video to learn about the CFPB and how we go to  work for you."

File: _includes/templates/nav/primary-nav.html [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+               list__horizontal
+               list-expanding">
+        {%- for nav_title, sub_nav_title, sub_nav_media, sub_nav_items in vars.nav_items %}
+            <li class="list_item list-expanding_list-item">
+                {%- if sub_nav_items|length > 0 %}
+                    {# TODO: Make this a <button>, it is not a real link. #}
+                    <a class="list_link list-expanding_trigger" href="#">
+                        {{ nav_title }}
+                    </a>
+                    {% import 'macros/sub-nav.html' as subNav %}
```
Comment (jimmynotjim) [inline _includes/templates/nav/primary-nav.html]: `sub-nav.html` -> `subNav` We typically haven't changed the name when importing a macro, any reason to here?

File: primary-nav-items.html [hunk 21]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (jimmynotjim) [inline primary-nav-items.html]: This file is a duplicate of `_includes/templates/nav/_vars-primary-nav.html`. We probably only need one right?

File: src/static/css/nav-primary.less [hunk 22]
```diff
@@ -0,0 +1,459 @@
...
+    }
+
+    .sub-nav_menu-list {
+        width: 100%;
+
+    }
+
+    .sub-nav_menu-list .list_link {
+        // Override standard visited link color.
+        .u-link__colors(@pacific, @pacific, @pacific-50, @pacific, @navy);
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: If we more directly targeted the changes on [L:69](https://github.com/cfpb/cfgov-refresh/pull/685/files#diff-326cef213344c8a38e386ce6019ac100R69) could we remove this since these are the defaults?

File: src/static/css/nav-primary.less [hunk 23]
```diff
@@ -0,0 +1,459 @@
...
+    }
+
+    .primary-nav_top-level-list.list-expanding {
+        // Nested list borders.
+        // Negative bottom margin to remove duplicate bottom border.
+        & .list-expanding {
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: None of the `&` with a space are needed because they blocks are already nested.

File: src/static/css/nav-primary.less [hunk 24]
```diff
@@ -0,0 +1,459 @@
...
+            padding: (@grid_gutter-width / 2)
+                     @grid_gutter-width
+                     (@grid_gutter-width / 2)
+                     (@grid_gutter-width / 2);
+            position: relative;
+            background: @gray-20;
+            font-size: unit(18px / @base-font-size-px, em);
+        }
+
+        & > .list_item > a:hover {
```
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: These link states could all be nested into the link block above for better code org.

File: src/static/css/nav-primary.less (hunk 11 above)
Comment (jimmynotjim) [inline src/static/css/nav-primary.less]: This extra indentation can still be removed

**Comment by jimmynotjim** [comment]
The question about the sub-nav naming when importing the macro is probably the only one that needs to be answered now. Everything else is more me making notes for when we clean this up (hopefully next month).

File: _includes/templates/nav/primary-nav.html (hunk 20 above)
Comment (sebworks) [inline _includes/templates/nav/primary-nav.html]: fixed in https://github.com/cfpb/cfgov-refresh/commit/9c9503576db1bcd264d330e79904be788043c18e

File: primary-nav-items.html (hunk 21 above)
Comment (sebworks) [inline primary-nav-items.html]: fixed in https://github.com/cfpb/cfgov-refresh/commit/9c9503576db1bcd264d330e79904be788043c18e

**Comment by jimmynotjim** [comment]
:+1: Let's get one more just to make sure :ok_hand:

**Comment by anselmbradford** [comment]
Is it intended to have the `About Overview` styled almost identical to the menu links but not be a link?

File: _includes/templates/nav/_vars-primary-nav.html [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+    nav_tbd,
+    nav_tbd,
+    nav_tbd,
+    ('About', 'About Overview', about_us_media , [
+        (
+            ('/the-bureau/', 'the-bureau', 'The Bureau'),
+            ('/budget/', 'budget', 'Budget and Strategy'),
+            ('/offices/payments-to-harmed-consumers/',
+             'payments-to-harmed-consumers',
+             'Payments to Harmed<br>Consumers'),
```
Comment (anselmbradford) [inline _includes/templates/nav/_vars-primary-nav.html]: May want to check with @duelj about the spacing when it falls to a newline ... the gap seems a little tall:

screen shot 2015-07-14 at 2 20 27 pm

Adjustments here can be in another PR I think.

File: _includes/templates/nav/about-us-media.html (hunk 19 above)
Comment (anselmbradford) [inline _includes/templates/nav/about-us-media.html]: cc @sarahsimpson09

File: _includes/templates/nav/primary-nav.html [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+            <li class="list_item list-expanding_list-item">
+                {%- if sub_nav_items|length > 0 %}
+                    {# TODO: Make this a <button>, it is not a real link. #}
+                    <a class="list_link list-expanding_trigger" href="#">
+                        {{ nav_title }}
+                    </a>
+                    {% import 'macros/sub-nav.html' as sub_nav %}
+                    {{ sub_nav.render(sub_nav_title, sub_nav_items, sub_nav_media) }}
+                {%- else %}
+                  <a class="list_link u-link__disabled">
```
Comment (anselmbradford) [inline _includes/templates/nav/primary-nav.html]: Minor: Indentation on else block inconsistent with if block.

File: _includes/templates/nav/_vars-primary-nav.html (hunk 25 above)
Comment (sebworks) [inline _includes/templates/nav/_vars-primary-nav.html]: @anselmbradford , I think it's a bit confusing but that is how it should be styled.

**Comment by anselmbradford** [comment]
Seems just to be minor cleanup from earlier comments. I'm fine with proceeding with merging. It would be great to capture the adjustments in issues. You may want to merge after the sprint PR so that we have a sprint to test it on flapjack.demo since it's a rather big feature.

**Comment by sebworks** [comment]
thanks, will wait till after the sprint PR is merged.

**Comment by sebworks** [comment]
@jimmynotjim , @anselmbradford  I've created issue https://github.com/cfpb/cfgov-refresh/issues/760  to capture any issues listed here. Please modify if I missed anything.

**Comment by anselmbradford** [comment]
:+1: that looks great!
